### PR TITLE
Implement fetching file URL's

### LINKED ISSUE #10166: Implement fetching file URLs
Code: `components/net/fetch/methods.rs`.
Spec: <https://fetch.spec.whatwg.org/#concept-basic-fetch>.

### PR DESCRIPTION
**Pull request by vramana**
closes #10166 

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10896)

### CODE REVIEW

**Comment by jdm** [comment]
This should use mime_guess::guess_mime_type for the MIME type.

**Comment by vramana** [comment]
@jdm Thanks. I was about to ask that in IRC.

**Comment by vramana** [comment]
@jdm Is it possible to write unit tests for this?

**Comment by KiChjang** [comment]
Yes, they are located at `tests/unit/net/fetch.rs`

**Comment by jdm** [comment]
Unit tests are in tests/unit/net/fetch.rs. You could use resources_dir_path to grab resources/sero.css and turn the result into a URL then fetch that.

File: components/net/fetch/methods.rs [hunk 1]
```diff
@@ -313,7 +316,31 @@ fn basic_fetch(request: Rc<Request>) -> Response {
             }
         },
 
-        "blob" | "file" | "ftp" => {
+        "file" => {
+            if *request.method.borrow() == Method::Get {
+                url.to_file_path().ok()
+                    .and_then(|file_path| {
```
Comment (KiChjang) [inline components/net/fetch/methods.rs]: nit: Wrong level of indentation.

File: components/net/fetch/methods.rs [hunk 2]
```diff
@@ -313,7 +316,31 @@ fn basic_fetch(request: Rc<Request>) -> Response {
             }
         },
 
-        "blob" | "file" | "ftp" => {
+        "file" => {
+            if *request.method.borrow() == Method::Get {
+                url.to_file_path().ok()
```
Comment (KiChjang) [inline components/net/fetch/methods.rs]: Using a match-expression here makes more sense.

File: components/net/fetch/methods.rs [hunk 3]
```diff
@@ -317,7 +320,32 @@ fn basic_fetch(request: Rc<Request>, cache: &mut CORSCache) -> Response {
...
+                            response.headers.set(ContentType(mime));
+                            response
+                        } else {
+                            Response::network_error()
+                        }
+                    },
+                    _ => Response::network_error()
+                }
+            } else {
+                Response::network_error()
```
Comment (vramana) [inline components/net/fetch/methods.rs]: @KiChjang I have used `match` now but there are too many `Response::network_error()`. Do you prefer this?? I feel that it's bit worse.

Comment (KiChjang) [inline components/net/fetch/methods.rs]: The previous version you had has a `None`. That is essentially a `Response::network_error()`  in disguise, which is my main issue.

File: components/net/fetch/methods.rs [hunk 4]
```diff
@@ -317,7 +320,32 @@ fn basic_fetch(request: Rc<Request>, cache: &mut CORSCache) -> Response {
             }
         },
 
-        "blob" | "file" | "ftp" => {
+        "file" => {
+            if *request.method.borrow() == Method::Get {
+                match url.to_file_path() {
+                    Ok(file_path) => {
+                        let file_path_dup = &file_path.to_owned();
+                        if let Some(mut file) = File::open(file_path).ok() {
```
Comment (KiChjang) [inline components/net/fetch/methods.rs]: I don't see the reason to convert `File::open(file_path)` to an `Option`.

Comment (KiChjang) [inline components/net/fetch/methods.rs]: Also, can this use `file_path.clone()` instead? That way you can get rid of `file_path_dup`.

Comment (KiChjang) [inline components/net/fetch/methods.rs]: We could also potentially use `.map_or` here, so the whole thing would look something like:
```rust
File::open(file_path.clone()).ok().map_or(Response::network_error(), |mut file| {
    let mut bytes = vec![];
    let _ = file.read_to_end(&mut bytes);
    let mime = guess_mime_type(file_path);

    let mut response = Response::new();
    *response.body.lock().unwrap() = ResponseBody::Done(bytes);
    response.headers.set(ContentType(mime));
    response
})

File: components/net/fetch/methods.rs (hunk 3 above)
Comment (vramana) [inline components/net/fetch/methods.rs]: Okay.

**Comment by vramana** [comment]
@KiChjang I have made the suggested changes.

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks!

**Comment by jdm** [comment]
@bors-servo: retry
Git

**Comment by KiChjang** [comment]
@bors-servo retry
* The annoying git

**Pull request merged**